llvm/lib/CodeGen/CodeGenPrepare.cpp

```diff
@@ -2298,8 +2298,29 @@ static bool despeculateCountZeros(IntrinsicInst *CountZeros,
   if (match(CountZeros->getOperand(1), m_One()))
     return false;
 
-  // If it's cheap to speculate, there's nothing to do.
   Type *Ty = CountZeros->getType();
+  EVT VTy = TLI->getValueType(*DL, Ty);
+
+  // do not despeculate if we have (ctlz (xor op -1)) if the operand is
```

Comment:
Should have IR test for this handling. The promotion check logic probably doesn't work correctly for vectors too

Reply:
Capitalize comment. Also I guess given the vector check above, the vector case doesn't matter?

Reply:
I'll capitalise the comment but I don't understand why the check for TypeAction/OperationAction is not right in the case of a vector type (although here it's valid b/c previous check ensuring it's not a vector type).

Reply:
I made a mistake though since I should have used `InstructionOpcodeToISD`.

Reply:
The vector case is skipped above, so I guess it doesn't matter.

The problem with checking the promotion logic like this, and in many other places, is there may be multiple type legalization steps before the operation is legal, so this isn't a 100% accurate way of determining the type this will ultimately use